### PR TITLE
add f2008 R1206 procedure-stmt support (closes issue 405)

### PR DESCRIPTION
**Pull request by rupertford**
Adds support for the F2008 R1206 procedure_stmt rule:
`procedure-stmt is [ MODULE ] PROCEDURE [ :: ] procedure-name-list`
which differs slightly from the F2003 R1206 rule by allowing optional '::':
`procedure-stmt is [ MODULE ] PROCEDURE procedure-name-list`

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review.

File: src/fparser/two/Fortran2008.py [hunk 1]
```diff
@@ -1578,6 +1579,54 @@ def match(string):
         )
 
 
+class Procedure_Stmt(Procedure_Stmt_2003):  # R1206
```
Comment (arporter) [inline src/fparser/two/Fortran2008.py:1582]: I note this file is now well over 1K lines. We really don't want to carry on towards the situation we have with `Fortran2003.py`. Perhaps we should have a new `Fortran2003` directory and break out the classes into separate files within it. If we agree, that can be a separate PR as it's going to be big and disruptive...

File: src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2023 Science and Technology Facilities Council
```
Comment (arporter) [inline src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py:1]: pylint says:

    test_procedure_stmt_r1206.py:93:40: W0612: Unused variable 'info' (unused-variable)
    test_procedure_stmt_r1206.py:42:0: W0611: Unused get_reader imported from fparser.api (unused-import)
    test_procedure_stmt_r1206.py:42:0: W0611: Unused walk imported from fparser.api (unused-import)

File: src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+"""
+
+import pytest
+from fparser.api import get_reader, walk
+from fparser.two.Fortran2008 import Procedure_Stmt
+from fparser.two.utils import NoMatchError
+
+
+@pytest.mark.usefixtures("f2008_create")
```
Comment (arporter) [inline src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py:47]: You don't need this `usefixtures` (or any of the others) because the `fortran2008/conftest.py` does this for you.

File: src/fparser/two/tests/test_fortran2003.py [hunk 4]
```diff
@@ -2964,6 +2964,9 @@ def test_procedure_stmt():  # R1206
     assert isinstance(obj, tcls), repr(obj)
     assert str(obj) == "MODULE PROCEDURE a, b"
 
+    with pytest.raises(NoMatchError):
```
Comment (arporter) [inline src/fparser/two/tests/test_fortran2003.py:2967]: Please add a comment to say that the double colon is only valid in Fortran2008 onwards.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for speedily fixing this. Just a couple of minor things to fix and then this can go on.
Please could you also update the introductory paragraph to fparser2 (in `fparser2.rst`) to mention this extension. Please could you also update it to mention `block` and `critical` which I forgot to do last time (mea culpa).

File: src/fparser/two/Fortran2008.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/Fortran2008.py:1582]: Yes I agree, if we can work out how to do it with the class generation code.

File: src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py (hunk 2 above)
Comment (rupertford) [reply src/fparser/two/tests/fortran2008/test_procedure_stmt_r1206.py:1]: Oops.

File: src/fparser/two/tests/test_fortran2003.py (hunk 4 above)
Comment (rupertford) [reply src/fparser/two/tests/test_fortran2003.py:2967]: Good idea.

**Comment by rupertford** [comment]
Ready for next review from @arporter. Further to our telco discussion I've changed from boolean's for the optional arguments to strings/None and re-ordered so the new optional arguments are at the end of items, which means that the first items entry is the same for both the 2003 and 2008 classes.

**Review by arporter — APPROVED** [review]
All requested changes have been made and the docs updated.
Docs build fine.
Will proceed to merge.
